pkg/kube/watcher/watcher.go

```diff
@@ -0,0 +1,94 @@
+package watcher
+
+import (
+	"fmt"
+
+	"github.com/jenkins-x/jx/pkg/log"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+// Watcher lists and watches resources
+type Watcher struct {
+	Namespace    string
+	ListOptions  metav1.ListOptions
+	HandlerFuncs HandlerFuncs
+	ListWatch    ListWatch
+}
+
+// HandlerFuncs is an adaptor to let you easily specify as many or
+// as few of the notification functions as you want
+type HandlerFuncs struct {
+	AddFunc    func(obj runtime.Object)
+	ModifyFunc func(obj runtime.Object)
+	DeleteFunc func(obj runtime.Object)
+}
+
+// ListWatch an interface for something that can list and watch resources
+type ListWatch interface {
+	List(ns string, listOptions metav1.ListOptions) ([]runtime.Object, error)
+	Watch(ns string, listOptions metav1.ListOptions) (watch.Interface, error)
+}
+
+// Run watches the resources until the connection is lost
+func (w *Watcher) Run() error {
+	log.Logger().Infof("watching and listing resources in namespace %s", w.Namespace)
+
+	watcher, err := w.ListWatch.Watch(w.Namespace, w.ListOptions)
+	if err != nil {
+		log.Logger().Fatalf("failed to watch resources in namespace %s due to %s", w.Namespace, err.Error())
+	}
+
+	resources, err := w.ListWatch.List(w.Namespace, w.ListOptions)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.Logger().Fatalf("failed to list resources in namespace %s due to %s", w.Namespace, err.Error())
```

Comment:
Is this mitigating an actual problem and how does a restart of the controller help?

Reply:
if the API server is down, we terminate the pod - so that it will keep retrying until a list/watch works

Reply:
I think this will crash the build controller POD completely making Kuberntes scheduler to restart the Pod and eventually fail based on the liveness probe setting when something is wrong in the cluster.

In the same time, there is an infinite for loop a the end of the build controller which keeps executing the Run if an error occurs passed this point. This looks like a second retrying strategy in case the code execution passes this point. 

I think some times, the result channel can be just closed without returning an error. From docs:
```
	// Returns a chan which will receive all the events. If an error occurs
	// or Stop() is called, this channel will be closed, in which case the
	// watch should be completely cleaned up.
```

I might be useful to have only one retrying mechanism in all use cases. I would rather propagate the error out of Run and let the outer for to decide based on error type how to handle the situation instead of using a `log.Fatalf`.

Reply:
we could switch the `log.Fatal` if we cannot List/Watch and return an error. The tricky decision is knowing when to terminate the container or not from outside